# Visualize the simulation results
# Libraries ----
suppressMessages(library(ggplot2))
suppressMessages(library(parallel))
suppressMessages(library(tidyr))
suppressMessages(library(dplyr))
suppressMessages(library(scales))
suppressMessages(library(data.table))
suppressMessages(library(ComplexHeatmap))
suppressMessages(library(circlize))
suppressMessages(library(cowplot))
suppressMessages(library(ggnewscale))
suppressMessages(library(ggExtra))

# Input from snakemake

# Path to the file containing the coverage of each positions in the concatenated contig from profiler
position_coverage_path <- snakemake@params[["position_coverage"]]
# Path to the file containing positions of simulated SNPs
snps_path <- snakemake@input[["mutations"]]
# Path to the gene gain log from simulation
gain_path <- snakemake@input[["gain"]]
# Path to the gene loss log from simulation
loss_path <- snakemake@input[["loss"]]
# Path to the file containing entropy value of every site in the concatenated contig
entropy_path <- snakemake@input[["weight_mutation"]]
# Path to the MGE data from profiler
mges_path <- snakemake@params[["mge_data"]]
# Path to the RDS which contains the bin info
chr_bins_path <- snakemake@params[["chromosomal_bins"]]
# Path to the SNP matrix generated by snp-dists
matrix_path <- snakemake@input[["snp_matrix"]]
# Path to directory where the output files will be saved
output_dir <- snakemake@params[["output_dir"]]

cat("\n")
cat("Chromosome bin file:", chr_bins_path, "\n")
cat("Simulation SNP file:", snps_path, "\n")
cat("MGEs data file:", mges_path, "\n")
cat("Gain log file:", gain_path, "\n")
cat("Loss log file:", loss_path, "\n")
cat("Entropy file:", entropy_path, "\n")
cat("SNP matrix:", matrix_path, "\n")
cat("Position coverage:", position_coverage_path, "\n")
cat("Output directory:", output_dir)
cat("\n")

# Function for visualization ----
simulation_visual <- function(output_dir,
                              snps_path,
                              mges_path,
                              position_coverage_path,
                              entropy_path,
                              chr_bins_path,
                              matrix_path,
                              gain_path,
                              loss_path){
  ## Load input ---- 
  setwd(dir = output_dir)
  snps_df <- read.csv(snps_path)
  entropy_df <- read.csv(entropy_path)
  entropy_df <- entropy_df[,c("profile_position","normalized_entropy")]
  chr_bins <- read.csv(chr_bins_path)
  mges_df <- readRDS(mges_path)
  gain_df <- read.csv(gain_path)
  loss_df <- read.csv(loss_path)
  snp_matrix <- read.delim(matrix_path,
                           header = TRUE,
                           row.names = 1)
  
  position_coverage <- readRDS(position_coverage_path)
  ## Coverage plot and Gene Gain/Loss Simulation plot ----
  coverage_plot_df <- position_coverage %>%
    mutate(cov_pct = 100*coverage/ max(position_coverage$coverage))
  
  
  coverage_plot <- ggplot() +
    # the rect showing the MGEs
    geom_rect(data = mges_df,
              aes(xmin = start,
                  xmax = end,
                  ymin = -Inf,
                  ymax = Inf),
              fill = "#41b6e6",
              color = "transparent",
              linewidth = 1,
              alpha = 0.65) +
    geom_point(data = coverage_plot_df,
               aes(x = position,
                   y = cov_pct),
               alpha = 0) + 
    geom_line(data = coverage_plot_df,
              aes(x = position,
                  y = cov_pct),
              linewidth = 0.1,
              color = "black",
              alpha = 0.85) + 
    scale_y_continuous(name = "% Coverage") +
    scale_x_continuous(expand = c(0,0)) + 
    theme(
      axis.title.y.left = element_blank(),
      axis.text.y.left = element_text(colour = "black",
                                      size = 15),
      axis.title.x = element_blank(),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      panel.grid = element_blank(),
      plot.background = element_rect(fill = "transparent"),
      panel.background = element_rect(fill = "transparent"),
      legend.position = "none",
      panel.border = element_rect(colour = "black",
                                  fill = NA,
                                  linewidth = 1)
    )
  # Get the data frame as the input for gain_loss_plot
  gain_loss_plot_df <- rbind(
    # Gain
    do.call(rbind,
            lapply(unique(gain_df$insert_bin),
                   function(bin_entry){
                     bin_id <- strsplit(as.character(bin_entry),split = "\\_")[[1]][1]
                     bin_start = chr_bins$start[chr_bins$bin_index == bin_id]
                     bin_end = chr_bins$end[chr_bins$bin_index == bin_id]
                     
                     data.frame(
                       category = "gain",
                       start = bin_start,
                       end = bin_end
                     )
                   })),
    # Loss
    do.call(rbind,
            lapply(unique(loss_df$loss_bin),
                   function(bin_entry){
                     bin_id <- strsplit(as.character(bin_entry),split = "\\_")[[1]][1]
                     bin_start = chr_bins$start[chr_bins$bin_index == bin_id]
                     bin_end = chr_bins$end[chr_bins$bin_index == bin_id]
                     
                     data.frame(
                       category = "loss",
                       start = bin_start,
                       end = bin_end
                     )
                   })))
  # gain_loss_plot shows the simulated gene gain and loss 
  gain_loss_plot <- ggplot() +
    # the rect showing gain events
    geom_rect(data = subset(gain_loss_plot_df,category == "gain"),
              aes(xmin = start,
                  xmax = end,
                  ymin = 0.5,
                  ymax = 1),
              fill = "#011F5B",
              color = "#011F5B",
              linewidth = 0.2,
              alpha = 0.75) +
    # the rect showing loss events
    geom_rect(data = subset(gain_loss_plot_df,category == "loss"),
              aes(xmin = start,
                  xmax = end,
                  ymin = 0,
                  ymax = 0.5),
              fill = "#990000",
              color = "#990000",
              linewidth = 0.2,
              alpha = 0.75) + 
    geom_hline(yintercept = 0.5, linewidth = 0.65, colour = "black") +
    scale_y_continuous(expand = c(0,0),
                       breaks = c(0.25, 0.75),
                       labels = c("Loss", "Gain")) + 
    scale_x_continuous(expand = c(0,0),
                       limits = c(0,max(coverage_plot_df$position))) + 
    theme(
      axis.title.y = element_text(size = 20,
                                  face = "bold"),
      axis.text.y = element_text(colour = "black",
                                 size = 15),
      axis.ticks.y = element_blank(),
      axis.title.x = element_blank(),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      panel.grid = element_blank(),
      plot.background = element_rect(fill = "transparent"),
      panel.background = element_rect(fill = "transparent"),
      legend.position = "none",
      panel.border = element_rect(colour = "black",
                                  fill = NA,
                                  linewidth = 1)
    )
  
  
  # Bin used for visualization but not for simulation
  # Only visualize the bins with simulated SNPs. Thus get the index for bins with snps
  simul_snps_plot_df <- rbindlist(lapply(seq_len(nrow(chr_bins)),function(row_idx){
    
    
    bin_idx <- chr_bins$bin_index[row_idx]
    bin_start <- chr_bins$start[row_idx]
    bin_end <- chr_bins$end[row_idx]
    bin_midpos <- mean(bin_start,bin_end)
    bin_snp_count <- sum(snps_df$profile_position >= bin_start &
                           snps_df$profile_position <= bin_end)
    
    return(
      data.table(
        bin_index = bin_idx,
        mid_pos = bin_midpos,
        snp_count = bin_snp_count
      )
    )
    
  }))
  
  # Plot of Simulated SNPs
  simul_snps_plot <- ggplot(simul_snps_plot_df,
                            aes(x = mid_pos,
                                y = snp_count)) +
    geom_col(position = "identity",
             fill = "#ed1f7f",
             color = "#ed1f7f",
             width = 0.1,
             alpha = 0.5) +
    scale_x_continuous(name = "Concatenated Contig Position (bp)",
                       limits = c(0,max(coverage_plot_df$position)),
                       expand = c(0,0)) +
    scale_y_continuous(name = "Simulated SNP",
                       expand = c(0,0)) +
    theme(axis.line.x = element_line(),
          axis.line.y = element_line(),
          axis.text.x = element_text(size = 15),
          axis.title.x = element_text(size = 20,
                                      face = "bold"),
          axis.title.y = element_blank(),
          axis.text.y = element_text(colour = "black",
                                     size = 15),
          plot.background = element_blank(),
          panel.background = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour = "black",
                                      fill = NA,
                                      linewidth = 1))
  
  # Plot for normalized entropy
  
  mut_prob_plot <- ggplot(entropy_df,
                            aes(x = profile_position,
                                y = normalized_entropy)) +
    geom_col(position = "identity",
             fill = "#91a01e",
             color = "#91a01e",
             width = 0.1,
             alpha = 0.5) +
    scale_x_continuous(name = "Concatenated Contig Position (bp)",
                       limits = c(0,max(coverage_plot_df$position)),
                       expand = c(0,0)) +
    scale_y_continuous(name = "% Mutation Probability",
                       expand = c(0,0),
                       labels = scales::label_scientific()) +
    theme(axis.line.x = element_line(),
          axis.line.y = element_line(),
          axis.text.x = element_blank(),
          axis.title.x = element_blank(),
          axis.title.y = element_blank(),
          axis.text.y = element_text(colour = "black",
                                           size = 15),
          plot.background = element_blank(),
          panel.background = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour = "black",
                                      fill = NA,
                                      linewidth = 1))
  
  combined_plot <- plot_grid(
    coverage_plot,
    gain_loss_plot,
    mut_prob_plot,
    simul_snps_plot,
    ncol = 1,
    align = "v",
    axis = "tblr",
    rel_heights = c(1,1,1.5,2))
  
  png(file=file.path(output_dir,"simulation_visualization.png"),
      width = 10000,
      height = 5000,
      res = 600)
  print(combined_plot)
  dev.off()
  
  # Visualize the snp distance heatmap
  
  snp_heatmap <- Heatmap(as.matrix(snp_matrix),
          show_column_names = FALSE,
          #column_names_side = "bottom", 
          #column_names_gp = gpar(fontsize = 5),
          #column_names_rot = 45,
          show_row_names = FALSE,
          #row_names_gp = gpar(fontsize = 5),
          show_heatmap_legend = TRUE,
          cluster_columns = TRUE,
          row_names_side = "left",
          show_row_dend = TRUE,
          show_column_dend = FALSE,
          row_dend_side = "left",
          clustering_method_rows = "complete",
          col = colorRamp2(c(min(snp_matrix, na.rm = TRUE), max(snp_matrix, na.rm = TRUE)), 
                           c("white", "#005587")),
          heatmap_legend_param = list(
            legend_width  = unit(0.5, "npc"),
            title = "SNP Distance",
            direction = "horizontal",
            title_position = "topcenter"
          ),
          width = unit(1, "snpc"),
          height = unit(1, "snpc"),
          use_raster = FALSE) %>%
    draw(heatmap_legend_side = "top")
  
  pdf(file=file.path(output_dir,"snp_matrix_visualization.pdf"),
      width=5,
      height=5)
  print(snp_heatmap)
  dev.off()

  # Remove the Rplots.pdf generated by default
  if(file.exists(file.path(output_dir,"Rplots.pdf"))){
    file.remove(file.path(output_dir,"Rplots.pdf"))
  }

}

# Execute Function ----
simulation_visual(output_dir = output_dir,
                  snps_path = snps_path,
                  mges_path = mges_path,
                  position_coverage_path = position_coverage_path,
                  entropy_path = entropy_path,
                  chr_bins_path = chr_bins_path,
                  matrix_path = matrix_path,
                  gain_path = gain_path,
                  loss_path = loss_path)
